File: Data Structures/queue/queue.py




class Queue(object):
    
    
    def __init__(self, capacity: int):
        self.__capacity__ = capacity
        self.__arr__ = [None for _ in range(capacity)]
        self.__head__ = 0
        self.__tail__ = 0
        self.__size__ = 0
        
        
    
    def enqueue(self, value: int):
        
        self.__arr__[self.__tail__] = value
        
        self.__tail__ += 1
        self.__tail__ %= self.__capacity__
        self.__size__ += 1
    
    
    
    def dequeue(self) -> int:
        
        if self.is_empty():
            raise Exception("Queue is empty")
        
        output = self.__arr__[self.__head__]
        
        self.__head__ += 1
        self.__head__ %= self.__capacity__
        self.__size__ -= 1
        
        return output
    
    
    
    def values(self) -> list:
        
        output = []
        
        h = self.__head__
        t = self.__tail__
        
        while h != t:
            output.append(self.__arr__[h])
            h = (h+1)%self.__capacity__
            
        return output
    
    
    
    def capacity(self) -> int:
        return self.__capacity__
    
    
    
    def empty_space(self) -> int:
        return self.size() - self.__len__()
    
    
    
    def is_empty(self) -> bool:
        return (self.__size__ == 0)
    
    
    
    def is_full(self) -> bool:
        return (self.__size__ == self.__capacity__)
        
        
        
    def __len__(self) -> int:
        return self.__size__
        
        
    
    
    
    
    
    
    
    
    
    
    
    
if __name__ == "__main__":
    
    print("\n\n ---------------- \n")
    print("a simple test:\n\n\n")
    
    
    # initialize a new queue
    q = Queue(1000)
    
    # add some values
    q.enqueue(1)
    q.enqueue(2)
    q.enqueue(3)
    q.enqueue(4)
    q.enqueue(5)
    q.enqueue(6)
    q.enqueue(7)
    q.enqueue(8)
    q.enqueue(9)
    q.enqueue(10)
    q.enqueue(11)
    
    
    print("stack: ", *(q.values()), '\n\n')
    
    
    q.enqueue(12)
    print("after enqueue value 12: ", *(q.values()), '\n')


    print("call dequeue function: ", q.dequeue(), '\n')

    print("after dequeue value 12: ", *(q.values()), '\n')
    
    
    
    
    
    
  
    print("\n\n ---------------- \n\n")
     
     

File: Data Structures/queue/queue.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <stdbool.h>



#define raiseERROR(x) {printf("\n\x1b[31m \berror:  \x1b[0m"#x"\n\n\n"); exit(EXIT_FAILURE);}
#define nl printf("\n")








/*      queue struct        */

typedef struct __QUEUE__
{
    int *arr;
    unsigned int capacity;

    int head;
    int tail;
    int size;

} queue;









/*      functions       */


queue *newQueue(int capacity)
{
    queue *output = (queue *) malloc(sizeof(queue));
    output->arr = (int *) malloc(capacity*sizeof(int));
    output->capacity = capacity;
    output->head = 0;
    output->tail = 0;
    output->size = 0;

    return output;
}


bool is_empty(queue *q)
{
    return (q->size == 0);
}


bool is_full(queue *q)
{
    return (q->size == q->capacity);
}


void enqueue(queue *q, int value)
{
    if (is_full(q))
        raiseERROR(Queue is full);

    q->arr[q->tail] = value;
    q->tail = (q->tail + 1) % q->capacity;
    q->size++;
}


int dequeue(queue *q)
{
    if (is_empty(q))
        raiseERROR(Queue is empty);

    int output = q->arr[q->head];

    q->head = (q->head + 1) % q->capacity;
    q->size--;

    return output;
}


int capacity(queue *q)
{
    return q->capacity;
}


int empty_space(queue *q)
{
    return (q->capacity - q->size);
}


void printqueue(queue *q)
{
    int h, t;
    h = q->head;
    t = q->tail;

    printf("queue:  ");
    while (h != t)
    {
        printf("%d, ", q->arr[h]);
        
        h++;
        h %= (q->capacity) -1;
    }
    printf("\b\b\n");

    return;
}







int main()
{
    printf("\n\n ---------------- \n\n");
    printf("a simple test:\n\n");

    
    // initialize a new queue
    queue *q = newQueue(100);

    // add some values
    enqueue(q, 1);
    enqueue(q, 2);
    enqueue(q, 3);
    enqueue(q, 4);
    enqueue(q, 5);
    enqueue(q, 6);
    enqueue(q, 7);
    enqueue(q, 8);
    enqueue(q, 9);
    enqueue(q, 10);
    enqueue(q, 11);
    enqueue(q, 12);

    
    printqueue(q);
    nl;


    enqueue(q, 14);
    printf("after push value 14:  \n"); 
    printqueue(q);
    nl;

    
    

    printf("call pop function:  %d\n", dequeue(q));
    nl;


    printf("after pop value 14:  "); printqueue(q);
    nl;




    printf("\n\n ---------------- \n\n\n");

    return 0;
}


